Fix missing venv path when running Pyright

krishnan-chandra opened the pull request:
This fix should address https://github.com/pantsbuild/pants/issues/19368 - I've tested it out locally on my own repo and haven't been able to reproduce the issue so far.

Relevant Slack threads:
https://pantsbuild.slack.com/archives/C046T6T9U/p1687366638498259
https://pantsbuild.slack.com/archives/C046T6T9U/p1688507180323199?thread_ts=1688480012.892699&cid=C046T6T9U

This is my first PR to Pants, so please let me know if there are any steps I forgot!

benjyw commented:
Thanks for tackling this @krishnan-chandra!

I do think John's suggestion over in https://github.com/pantsbuild/pants/issues/19368#issuecomment-1638432081 is better than mine though - less finicky and brittle.

That solution involves running a no-op process in the venv. You do something like:

```
await Get(
        ProcessResult,
        VenvPexProcess(
            requirements_venv_pex,
            description="Force venv to materialize",
            argv=["-c", "''"],
        )
    )
```

benjyw commented:
Thanks @krishnan-chandra. Can you confirm that the problem reproduces without this latest patch and does not reproduce with it?

krishnan-chandra commented:
I can indeed! Here are two screenshots:

### Before Patch

<img width="961" alt="Screenshot 2023-07-24 at 4 19 53 PM" src="https://github.com/pantsbuild/pants/assets/1229365/6bb6225c-0c7a-4af4-babb-921ee5ea9a23">

### After Patch

<img width="962" alt="Screenshot 2023-07-24 at 4 20 55 PM" src="https://github.com/pantsbuild/pants/assets/1229365/07d3af9e-1dc6-4fa7-bb4a-7a1e65e67483">

krishnan-chandra commented:
For testing, I was thinking about how best to create a test for this case. I am able to reproduce it by deleting the full cache between test runs:

```bash
rm -rf ~/.cache/pants/named_caches/pex_root
```

This seems quite intense for a test though, so I'm wondering if there is a way to delete the path of the requirements venv between test runs. I couldn't figure out how to do that using the test harnesses in https://github.com/pantsbuild/pants/blob/main/src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py, but I'm sure I'm missing something on the internals there.

krishnan-chandra commented:
Thanks! I'll take a look at the examples and try to add those tests. In the meantime, are there any other changes that you would like me to make / is there anything else you need before approving the PR?

benjyw commented:
No, I think this looks good! And thanks for confirming that it has the desired effect. Just that test (and verifying that the test fails without the patch) and we're good!

krishnan-chandra commented:
I added a test - it passes, but I can't get it to fail if I cherry-pick the test onto a branch without the fix. Is there anything obvious that I'm doing wrong there / does the named caches dir need to be passed somewhere else?

benjyw commented on src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:
```diff
@@ -426,3 +431,54 @@ def assert_partition(
+            )
+        )
+
+        def run_partition_test():
```
I don't think this is testing the right thing. This is just generating the pyright input partitions and checking that they contain what you'd expect. But you want to test that actually running Pyright works under the relevant conditions. 

Look at the tests that call `run_pyright()` for guidance.

krishnan-chandra replied:
Good point. I changed the test to actually run Pyright and clear the cache in between, but I think the examples might be a bit too simple and I'm not sure how to force the test to break without introducing some third-party imports.

krishnan-chandra commented:
I've pushed up what I think should be a better test - but again, I can't get this one to fail consistently even though I can definitely get Pants to fail when running against my own monorepo.

benjyw commented:
> I've pushed up what I think should be a better test - but again, I can't get this one to fail consistently even though I can definitely get Pants to fail when running against my own monorepo.

Hmm, this would be absolutely key. If the test doesn't fail without the change then it is not really testing what we think it is...

krishnan-chandra commented:
Definitely agreed there - I'm not super familiar with Pants internals so it's been a bit challenging to chase down the source of the problem. I'll keep pushing on it. In the meantime, if someone more familiar with the Pyright integration has suggestions, I am all ears :)

benjyw commented:
I'll take a look when I get a moment.

benjyw commented:
OK, so one thing complicating this test is that it's too simple: Specifically, it doesn't actually require any venv to be present in order to pass. And so it passes regardless of the presence of absence of a venv (the --venvPath not existing is just a pyright warning, not an error).

So the next step here is to make the test sufficiently complex to actually require some 3rdparty dep to be present in order to pass.

benjyw commented:
Hi @krishnan-chandra , have you had a chance to look at this test again?

krishnan-chandra commented:
I did revisit it, and I tried to add some third-party imports to force the venv to materialize. One thing I couldn't quite work out is where to look for the output. Here is my test output:

**Command:**

```bash
./pants --keep-sandboxes=always test src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py -- -k test_passing_cache_clear
```

**Output:**

```
19:21:17.08 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-AivibN for Searching for `bash` on PATH=/usr/bin:/bin:/usr/local/bin
19:21:17.26 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-G5A9c6 for Test binary /bin/bash.
19:21:17.31 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-BnaswX for Extracting plugin locations
19:21:18.92 [INFO] Initializing scheduler...
19:21:32.03 [INFO] Scheduler initialized.
19:21:33.97 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-kwvimH for Find interpreter for constraints: CPython==3.10.*
19:21:35.77 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-7NoP1T for Searching for `bash` on PATH=/usr/bin:/bin:/usr/local/bin
19:21:35.88 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-Rr5xKf for Test binary /bin/bash.
19:21:35.94 [INFO] Preserving local process execution dir /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-L322Gu for Run Pytest for src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:tests
19:21:49.25 [ERROR] Completed: Run Pytest - src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:tests - failed (exit code 1).
============================= test session starts ==============================
collecting ... collected 14 items / 13 deselected / 1 selected

src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py::test_passing_cache_clear FAILED [100%]

=================================== FAILURES ===================================
___________________________ test_passing_cache_clear ___________________________

rule_runner = RuleRunner(build_root=/private/tmp/_BUILD_ROOToahgp6be)

    def test_passing_cache_clear(rule_runner: PythonRuleRunner) -> None:
        # Ensure that the requirements venv must be created, by adding in a third-party
        # requirement to the test code.
        rule_runner.write_files({
            "BUILD": "python_requirement(name='more-itertools', requirements=['more-itertools==8.4.0'])",
            f"{PACKAGE}/f.py": dedent(
                """\
                from more_itertools import flatten

                assert flatten(42) == [4, 2]
                """
            ),
            f"{PACKAGE}/BUILD": "python_sources()",
        })
        print("Rule runner is running from directory: ", rule_runner.build_root)
        with temporary_dir() as named_caches:
            print("Temporary cache dir located at: ", named_caches)
            tgt = rule_runner.get_target(Address(PACKAGE, relative_file_path="f.py"))
            result = run_pyright(rule_runner, [tgt], extra_args=[f"--named-caches-dir={named_caches}"])
            assert len(result) == 1
>           assert result[0].exit_code == 0
E           assert 1 == 0
E             +1
E             -0

src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:163: AssertionError
----------------------------- Captured stdout call -----------------------------
Rule runner is running from directory:  /private/tmp/_BUILD_ROOToahgp6be
Temporary cache dir located at:  /tmp/tmpf9u5rvy1
- generated xml file: /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-L322Gu/src.python.pants.backend.python.typecheck.pyright.rules_integration_test.py.tests.xml -
```


This makes sense as I largely got the implementation from `test_thirdparty_dependency`, which is supposed to fail. What's weird though is that I can't find either of these directories even though I'm running with `--keep-sandboxes=always`:

`/private/tmp/_BUILD_ROOToahgp6be`
`/tmp/tmpf9u5rvy1`


And the following command does not

benjyw commented:
`temporary_dir()` is a context manager that deletes the tmpdir when the context ends (i.e., outside the `with` block). You can temporarily use `temporary_dir(cleanup=False)` to prevent that, but you do want the test to clean up after itself. This means that you need to do any checking of that dir inside the `with` block. This tmpdir is separate from any process sandbox.

The `/private/tmp/_BUILD_ROOToahgp6be` is also a tmpdir, created somewhere in run_pyright's callstack, so, again, is gone by the time you get to inspect it.

Does that make sense? Is that helpful?

benjyw commented:
PS sorry for the late reply. Summer, you know...

krishnan-chandra commented:
No worries at all! And sorry for the delay on my end too - I haven't had much time to look at this in the last few days but should have more time next week.

benjyw commented:
FWIW it seems like you're pretty close. Just need to nail down this test quirk, which admittedly is not simple.

krishnan-chandra commented:
Okay, I think I made some good progress here! What I had to do to debug the test:

1. Set `preserve_tmpdirs = True` when constructing the `PythonRuleRunner` object, which enabled me to retain and inspect the temporary directories created by the test.
2. Try to see what exactly gets put inside the sandbox directory that causes the test to fail.


When Pyright fails (inside the `with temporary_dir()` block), here is what I get:

```shell
Stdout:
 /private/tmp/RuleRunner.zes6l3vi/pants-sandbox-n0LG2F/src/py/project/f.py
  /private/tmp/RuleRunner.zes6l3vi/pants-sandbox-n0LG2F/src/py/project/f.py:1:6 - error: Import "more_itertools" could not be resolved (reportMissingImports)
1 error, 0 warnings, 0 informations

Stderr:
 venvPath /private/tmp/tmpm2i240me/pex_root is not a valid directory.
venv venvs/afe024b06a00fd7c8bc9445acb5d424bbcaec589/44edc6b3e5a957c5c1fc1b742988f030cd45baa2 subdirectory not found in venv path /private/tmp/tmpm2i240me/pex_root.
```

(Interestingly, I could not find the string `venvPath` or `venv path` anywhere in the Pants or Pex repos so it seems that this error is coming from somewhere else).

When I inspect the sandbox created by the RuleRunner, it looks like this:

```
> ls -ha /private/tmp/RuleRunner.zes6l3vi/pants-sandbox-n0LG2F/

.
..
._corepack_home
._npm
__node
__run.sh
_binary_shims_cdd60ac6ea9db484515d1bc36e0a9b992c89650ef4e5b32a200da6e9ae172946
pyrightconfig.json
requirements.pex
requirements_venv.pex
requirements_venv.pex_bin_python_shim.sh
requirements_venv.pex_pex_shim.sh
src
```

But for other sandboxes, I see that there is a cache directory:

```
ls -ha /System/Volumes/Data/private/tmp/RuleRunner.q1h3h5pd/pants-sandbox-zUPGQG/
.                        .cache                   .tmp                     pex                      source_files
..                       .python-build-standalone __run.sh                 requirements.pex
```


So I'm trying to figure out how to force delete the cache directory from the sandbox after the first run and then run it again instead of creating a separate temp directory. Does that make sense / is that the right approach?

---

Sorry for the wall of text here. I feel like I am very close to solving this one, but am missing some crucial piece of information that would connect the dots.

benjyw commented:
I think you're on the right track. The lack of a `.cache` directory (or, more specifically, a symlink to a venv somewhere inside it) is precisely what triggers the error, IIUC. So you should see that `.cache` directory only on success. Are you not seeing it in that case?

krishnan-chandra commented:
I think I got this one now - updated the tests to ensure that the original cache directory works both before and after trying a temp directory.

When I preserve the temp directory, no cache shows up but I think that is to be expected.

LMK if there's anything else you need here! Otherwise would love to kick off a test run and see if this works properly.

benjyw commented on src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:
```diff
@@ -287,6 +288,52 @@ def test_skip(rule_runner: PythonRuleRunner) -> None:
+        rule_runner.options_bootstrapper.bootstrap_options.for_global_scope().named_caches_dir
+    )
+
+    # Should now fail with a different cache directory, since there is no venv
```
I don't understand this. Isn't the whole point of this change that we are not relying on the venv existing, and we will force-create it if needed?

benjyw replied:
The right thing to simulate is deleting the named_cache, and then rerunning *without* changing any process inputs.

benjyw replied:
And you want to manually verify that this fails if you comment out the change in rules.py, but passes if you leave it in.

krishnan-chandra replied:
Understood on the last part - I have a separate branch that contains this test, but without the change in rules.py so I'm able to test between the two. Will rework the rest to adhere to what you mentioned.

krishnan-chandra commented on src/python/pants/backend/python/typecheck/pyright/rules.py:
```diff
@@ -194,6 +194,16 @@ async def pyright_typecheck_partition(
         ),
     )
 
+    # Force the requirements venv to materialize always by running a no-op.
```
I understood the first sentence, but I don't think I fully understood this one:

> So this needs `cache_scope=ProcessCacheScope.PER_SESSION`.

I see that the definition says this:

```
# Will run once per Session, i.e. once per run of Pants. This happens because the engine
# de-duplicates identical work; the process is neither memoized in memory nor cached to disk.
PER_SESSION = "per_session"
```

Is the venv cached per-session as well? What am I missing?

benjyw replied:
This is absolutely the key point, and it's important to clarify so we're all certain about the fix:

There are two caches in play here: The on-disk LMDB store which caches process executions and the on-disk venv cache, aka "named-caches". (There is also an in-memory cache of rule executions, but that is not impinging on the issue here.)

When running a pex process, the relevant underlying venv is either retrieved from cache, or created (and cached). However PyRight is special - it's a NodeJS binary that expects to operate against a Python virtualenv. So PyRight is *not* a pex process, and so doesn't create the venv, but assumes it was previously created. And without caches, it would be created on line 187, by the process that turns a PexRequest into a VenvPex. However, what if that process is cached in the LMDB store? It won't re-run. And so nothing ensures that the venv exists in the named-caches. If it does, we're fine, but if not, we blow up.

So the naive solution is to force that process to run every time, ignoring the cache, by setting cache_scope. But, that would potentially be expensive. So the more clever solution is the one you're implementing - run a separate no-op process that just ensures that the venv exists (and is cached), and then returns without doing anything, and so is fast. But we still have to force it to run every time! That is why it exists. 

Does that make sense?


And of course the test has to reflect this logic.

krishnan-chandra replied:
Absolutely makes sense! And you unlocked some helpful knowledge for me on how to debug the rest of the failures in the test too. 

Thanks for the very clear and comprehensive explanation!

krishnan-chandra commented:
Fixed up the test format to what it was before (run Pyright, clear cache, then run it again with the same inputs).

Interestingly, the issue I'm seeing now is that the `pex_root` directory under the cache dir doesn't get created, and the stderr of the test output looks like this:

```
Stderr:
 venvPath /private/tmp/tmpkzy7tf8g/pex_root is not a valid directory.
venv venvs/afe024b06a00fd7c8bc9445acb5d424bbcaec589/44edc6b3e5a957c5c1fc1b742988f030cd45baa2 subdirectory not found in venv path /private/tmp/tmpkzy7tf8g/pex_root.
```

In the most recent diff I added a change that would create the `pex_root` directory under the cache directory manually as part of the test. That removed the first line of the log from above (`venvPath /private/tmp/tmpkzy7tf8g/pex_root is not a valid directory.`).

But the other log remains where the venv subdirectory cannot be found under `pex_root`. So I have to assume that Pants is creating this venv somewhere else - even when I set `cleanup=False` for the temporary dir, the venv doesn't show up underneath it.

krishnan-chandra commented on src/python/pants/backend/python/typecheck/pyright/rules.py:
```diff
@@ -184,13 +190,28 @@ async def pyright_typecheck_partition(
-            internal_only=True,
-            pex_path=[requirements_pex],
-            interpreter_constraints=partition.interpreter_constraints,
+        VenvPexRequest(requirements_pex_request, complete_pex_env),
```
This change appears to solve the issue mentioned in https://github.com/pantsbuild/pants/pull/19430#issuecomment-1729763202 - here is my explanation of why (**please correct me if any of this is wrong**):

1. Previously, the `Get` took in a `PexRequest` as input and was set to return a `VenvPex` as output.
2. Pants used [this rule](https://github.com/pantsbuild/pants/blob/main/src/python/pants/backend/python/util_rules/pex.py#L989-L994) to convert the `PexRequest` into a `VenvPexRequest`.
3. Because the rule calls `pex_environment.in_sandbox` instead of `pex_environment.in_workspace`, the temporary named cache directory wasn't getting passed through correctly.
4. Now that we have moved up the declaration of `complete_pex_env` to be usable here, the requirements venv will actually get created under the correct cache directory as compared to before.

benjyw replied:
For context: the in_sandbox cache is just a symlink to the underlying cache that in_workspace also uses, which is `~/.cache/pants/named_caches/pex_root` by default. So the code is was correct for production use. But...

I patched in your change and debugged it, and what's happening is that the test infrastructure sets the named_caches_dir itself, to the "real" location of the enclosing Pants run (that `~/.cache/pants/named_caches/pex_root`). The `--named-caches-dir` in extra_args isn't consulted for this when the rule is run, as it's set too late (the named_caches_dir is captured at RuleRunner creation time). So without moving up the setting of complete_pex_env,  `{cache_dir}/pex_root` is never created, but *is* then consulted by pyright, hence the failures.

But when you move up the setting of `complete_pex_env`, it forces a re-consulting of the named_caches_dir option, which now does contain the tmpdir location, and things work as expected.

So the previous code was fine for production use, but it did not support the test. 

The real solution here is to fix the RuleRunner, but that's a bigger thing, and you've done more than enough hacking on this problem already :)

So I think this is good to go, with a comment in the code explaining that the complete_pex_env trick is needed just for the test, and opening an issue (assigned to me) about fixing the RuleRunner. 

Sound OK?

krishnan-chandra replied:
Absolutely - that's fine with me! Thanks so much for your grace and patience while I worked through fixing this :)

krishnan-chandra replied:
Issue created at #19946 - I can't assign it to you, but feel free to edit and assign as needed.

krishnan-chandra commented on src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:
```diff
@@ -287,6 +289,43 @@ def test_skip(rule_runner: PythonRuleRunner) -> None:
     assert not result
 
 
+def test_passing_cache_clear(rule_runner: PythonRuleRunner) -> None:
```
Note - I have a branch up called [`pyright-test-only`](https://github.com/krishnan-chandra/pants/tree/pyright-test-only) that contains only these test changes without the changes in `rules.py`.

Below are test outputs from my terminal for both branches:

**Passing:**

```bash
~/Projects/pants git:(pyright-venv-fix)

./pants test ./src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py -- -k test_passing_cache_clear
08:25:58.69 [INFO] Completed: Run Pytest - src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:tests - succeeded.

✓ src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:tests succeeded in 19.79s.
```

**Failing:**

```bash
~/Projects/pants git:(pyright-test-only)

./pants test ./src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py -- -k test_passing_cache_clear
There is no pantsd metadata at /Users/krishnanchandra/Projects/pants/.pants.d/pids/460a4b48a969/pantsd.
08:27:05.20 [INFO] Initializing scheduler...
08:27:18.04 [INFO] Scheduler initialized.
08:27:37.63 [ERROR] Completed: Run Pytest - src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:tests - failed (exit code 1).
============================= test session starts ==============================
collecting ... collected 14 items / 13 deselected / 1 selected

src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py::test_passing_cache_clear FAILED [100%]

=================================== FAILURES ===================================
___________________________ test_passing_cache_clear ___________________________

rule_runner = RuleRunner(build_root=/private/tmp/_BUILD_ROOTx4cb7qsb)

    def test_passing_cache_clear(rule_runner: PythonRuleRunner) -> None:
        # Ensure that the requirements venv must be created, by adding in a third-party
        # requirement to the test code.
        rule_runner.write_files(
            {
                "BUILD": "python_requirement(name='more-itertools', requirements=['more-itertools==8.4.0'])",
                f"{PACKAGE}/f.py": dedent(
                    """\
                from more_itertools import is_sorted

                assert is_sorted([1, 2, 3]) is True
                """
                ),
                f"{PACKAGE}/BUILD": "python_sources()",
            }
        )
        tgt = rule_runner.get_target(Address(PACKAGE, relative_file_path="f.py"))

        # On the first run, it should work as advertised with no modifications.
        with temporary_dir() as cache_dir:
            result = run_pyright(rule_runner, [tgt], extra_args=[f"--named-caches-dir={cache_dir}"])
            assert len(result) == 1
>           assert result[0].exit_code == 0
E           assert 1 == 0
E             +1
E             -0

src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:314: AssertionError
- generated xml file: /private/var/folders/0w/zhpm9fs12zzcmt3m5799xb9c0000gn/T/pants-sandbox-h7VtEl/src.python.pants.backend.python.typecheck.pyright.rules_integration_test.py.tests.xml -
=========================== short test summary info ============================
FAILED src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py::test_passing_cache_clear
====================== 1 failed, 13 deselected in 14.97s =======================



✕ src/python/pants/backend/python/typecheck/pyright/rules_integration_test.py:tests failed in 15.48s.
```

krishnan-chandra reviewed (commented):
Okay, should be much better now! Test flow should be relatively simple and I've included a new branch that contains only the test without the associated rules change. I can confirm that the test fails without the associated rules change.

benjyw reviewed (commented):
Thanks for hanging in here! I appreciate your persistence in getting this right.

This is good to go, but see below for a fuller explanation of what's going on and the last thing to do to get this over the finish line (basically a comment and an issue).

benjyw commented:
Thanks for your diligence here @krishnan-chandra !
